### ISSUE TITLE
Implement cache to work with job files

### ISSUE THREAD

**Issue by benedictpaten** [BD2KGenomics/toil#376]
Depends on #375 working

**Comment by benedictpaten** [comment]
Job files are not immutable. I spent a *long, long* time wondering how to make them immutable and concluded while it had some awesome benefits, it had some major downsides and would take a long time to implement. So, to make them work with the proposed cache I suggest the following algorithm:

- Let rootJob be the first job run by a worker.
- An "updatedJobs" list of all job/cached files updated/created during the run is maintained. 
- A "deletedJobs" list of all job files deleted during the run is maintained. 
- During Target._serialiseJobGraph() the "filesToDelete" of rootJob is set equal to "updatedJobs" and the rootJob is serialised to the disk. 
- Any other changes to rootJob are deferred until the end of the worker
- At the end of the worker, rootJobs "filesToDelete" is set to empty and and changes to the rootJob are made and the rootJob updated on disk
- Any jobs in deletedJobs are removed from the jobStore.

**Comment by benedictpaten** [comment]
I'm still waiting for my PR to go in, but I've started getting the middle cache layer done. I won't push anything till I see the other PR go in.
